src/util/connection_pool.rs

```diff
@@ -134,25 +160,47 @@ impl Context {
     ) {
         let context = self;
 
+        let context2 = context.clone();
+        let conn_fut = async move {
+            let conn = context2
+                .endpoint
+                .connect(node_id, &context2.alpn)
+                .await
+                .map_err(PoolConnectError::from)?;
+            if let Some(on_connect) = &context2.options.on_connected {
+                on_connect(&context2.endpoint, &conn)
+                    .await
+                    .map_err(PoolConnectError::from)?;
+            }
+            Result::<Connection, PoolConnectError>::Ok(conn)
+        };
+
         // Connect to the node
-        let state = context
-            .endpoint
-            .connect(node_id, &context.alpn)
+        let state = conn_fut
             .timeout(context.options.connect_timeout)
             .await
             .map_err(|_| PoolConnectError::Timeout)
-            .and_then(|r| r.map_err(PoolConnectError::from));
-        if let Err(e) = &state {
-            debug!(%node_id, "Failed to connect {e:?}, requesting shutdown");
-            if context.owner.close(node_id).await.is_err() {
-                return;
+            .and_then(|r| r);
+        let conn_close = match &state {
+            Ok(conn) => {
+                let conn = conn.clone();
+                MaybeFuture::Some(async move { conn.closed().await })
             }
-        }
+            Err(e) => {
+                debug!(%node_id, "Failed to connect {e:?}, requesting shutdown");
+                tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
Why do we sleep here?

Reply:
I haven't got the slightest idea. Probably wanted to see if some test fails, and forgot to remove. Thanks for noticing.